secrets/store.go

```diff
@@ -59,62 +58,63 @@ func (s *Store) parser(r io.Reader) (interface{}, error) {
 		return nil, err
 	}
 
-	if s.secretHandlerFunc != nil {
-		s.secretHandlerFunc(secrets)
-	}
+	s.secretHandlerFunc(secrets)
 
 	return secrets, nil
 }
 
-// SecretHandler creates the middleware chain.
+// secretHandler creates the middleware chain.
 func (s *Store) secretHandler(middlewares ...SecretMiddleware) {
-	if s.secretHandlerFunc == nil {
-		s.secretHandlerFunc = nopSecretHandlerFunc
-	}
-
 	for _, m := range middlewares {
 		s.secretHandlerFunc = m(s.secretHandlerFunc)
 	}
 }
 
-// GetSimpleSecret loads secrets from watcher, and fetches a simple secret from secrets
-func (s *Store) GetSimpleSecret(path string) (SimpleSecret, error) {
-	secrets, ok := s.watcher.Get().(*Secrets)
-	if !ok {
-		return SimpleSecret{}, fmt.Errorf("unexpected type %T", secrets)
-	}
+func (s *Store) getSecrets() *Secrets {
+	return s.watcher.Get().(*Secrets)
+}
 
-	return secrets.GetSimpleSecret(path)
+// Stop stops the underlying filewatcher and release associated resources.
+//
+// After Stop is called, you won't get any updates to the secret file,
+// but can still access the secrets as they were before Stop is called.
+//
+// It's OK to call Stop multiple times. Calls after the first one are no-ops.
+func (s *Store) Stop() {
+	s.watcher.Stop()
 }
 
-// GetVersionedSecret loads secrets from watcher, and fetches a versioned secret from secrets
-func (s *Store) GetVersionedSecret(path string) (VersionedSecret, error) {
-	secrets, ok := s.watcher.Get().(*Secrets)
-	if !ok {
-		return VersionedSecret{}, fmt.Errorf("unexpected type %T", secrets)
-	}
+// AddMiddlewares registers new middlewares to the store.
+//
+// Every AddMiddlewares call will cause all already registered middlewares to be
+// called again with the latest data.
+//
+// AddMiddlewares call is not thread-safe, it should not be called concurrently.
+func (s *Store) AddMiddlewares(middlewares ...SecretMiddleware) {
+	s.secretHandler(middlewares...)
+	s.secretHandlerFunc(s.getSecrets())
```

Comment:
As commented this would also trigger existing middlewares. What do you think have this function accepts  `SecretHandlerFunc`, call them and wrap them as `SecretMiddleware`?

https://github.com/reddit/baseplate.go/pull/71/files#diff-1a3fa2e612f9f352390e03634cc256d0R82

Reply:
triggering existing middlewares shouldn't be a problem, because we are triggering them with the same secret data.

If we want to accept `SecretHandlerFunc` instead of `SecretMiddleware` as the arg, we should make it consistent and also `SecretMiddleware` in `NewStore` as well (and not export `SecretMiddleware` at all).